cardano-node/src/Cardano/CLI/Delegation/PoolRegistry.hs

```diff
@@ -0,0 +1,301 @@
+
+
+
+module Cardano.CLI.Delegation.PoolRegistry
+  ( Cost
+  , Id
+  , Margin
+  , PledgeAddress
+  , RegistryRoot(..)
+  , Submission
+  , SubmissionFile(..)
+  , Ticker
+  , paDataPart
+  , paHumanReadable
+  , sCost
+  , sHomepage
+  , sId
+  , sMargin
+  , sPledgeAddress
+  , sTicker
+  , unCost
+  , unId
+  , unMargin
+  , unTicker
+  , validateRegistrySubmission
+  )
+where
+
+import           Prelude (String, fail, show)
+import           Cardano.Prelude hiding (option, show, trace, (%), (<.>))
+
+import qualified Codec.Binary.Bech32 as Bech32
+import           Data.Aeson (FromJSON(..), Value(..), (.:?), (.:))
+import qualified Data.Aeson as AE
+import qualified Data.Aeson.Types as AE
+import qualified Data.Aeson.Internal as AE
+import           Data.Aeson.Internal ((<?>))
+import           Data.Bifunctor (bimap)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.Char as Char
+import qualified Data.HashMap.Lazy as HMap
+import qualified Data.List as List
+import qualified Data.Set as Set
+import qualified Data.Scientific as Sci
+import qualified Data.Text as Text
+import           Data.Text (pack, unpack)
+import qualified Network.URI as URI
+
+import qualified System.Directory as Dir
+import qualified System.FilePath as Path
+import           System.FilePath ((</>), (<.>))
+
+import           Cardano.CLI.Ops
+
+--------------------------------------------------------------------------------
+-- * Exported API
+--
+newtype RegistryRoot =
+  RegistryRoot FilePath
+  deriving (Eq, Ord, Show, IsString)
+
+newtype SubmissionFile =
+  SubmissionFile FilePath
+  deriving (Eq, Ord, Show, IsString)
+
+data Submission = Submission
+  { sId :: !Id
+  , sTicker :: !Ticker
+  , sCost :: !Cost
+    -- | Absolute URI.
+  , sHomepage :: !URI.URI
+  , sMargin :: !Margin
+  , sPledgeAddress :: Maybe PledgeAddress
+  }
+
+-- | Stake pool ID, also its 32-byte public key.
+newtype Id = Id { unId :: Text }
+  deriving (Eq)
+
+-- | Stake pool ticker, a 3-4 all-ASCII-uppercase character ID.
+newtype Ticker = Ticker { unTicker :: Text }
+  deriving (Eq, Ord, Show)
+
+-- | Non-negative Scientific.
+newtype Cost = Cost { unCost :: Sci.Scientific }
+
+-- | Rational in range [0, 1]
+newtype Margin = Margin { unMargin :: Sci.Scientific }
+
+-- | Bech32-encoded address.
+data PledgeAddress = PledgeAddress
+  { paHumanReadable :: Bech32.HumanReadablePart
+  , paDataPart :: Bech32.DataPart
+  }
+
+-- | Given a stake pool registry root and a submission file,
+--   perform full validation of relative file structure
+--   & JSON content of the submission.
+validateRegistrySubmission
+  :: RegistryRoot
+  -> SubmissionFile
+  -> ExceptT CliError IO Submission
+validateRegistrySubmission (RegistryRoot root) (SubmissionFile fp) = do
+  -- Directory structure/access checks.
+  --
+  afp <- lift $ Dir.makeAbsolute fp
+  unlessM (lift $ Dir.doesFileExist afp) $
+    checks ["Submission doesn't exist: " <> pack afp]
```

Comment:
Eg: 
```
-     checks ["Submission doesn't exist: " <> pack afp]
+     left $ PoolRegistryValidationErrors fp ["Submission doesn't exist: " <> pack afp]
```

Reply:
You can maybe even get more specific with your error type `SubmissionDoesNotExist $ pack afp`. Up to you :).

Reply:
Yes, I was weighing that option, indeed..